binding/rust/src/rhino.rs

```diff
@@ -0,0 +1,543 @@
+
+            let pv_rhino_free_slots_and_values: Symbol<PvRhinoFreeSlotsAndValuesFn> =
+                load_library_fn!(lib, b"pv_rhino_free_slots_and_values");
+
+            let pv_rhino_reset: Symbol<PvRhinoResetFn> = load_library_fn!(lib, b"pv_rhino_reset");
+
+            let pv_rhino_context_info: Symbol<PvRhinoContextInfoFn> =
+                load_library_fn!(lib, b"pv_rhino_context_info");
+
+            let pv_sample_rate: Symbol<PvSampleRateFn> = load_library_fn!(lib, b"pv_sample_rate");
+
+            let pv_rhino_frame_length: Symbol<PvRhinoFrameLengthFn> =
+                load_library_fn!(lib, b"pv_rhino_frame_length");
+
+            let pv_rhino_version: Symbol<PvRhinoVersionFn> =
+                load_library_fn!(lib, b"pv_rhino_version");
+
+            let sample_rate = pv_sample_rate();
+            let frame_length = pv_rhino_frame_length();
+            let version = match CStr::from_ptr(pv_rhino_version()).to_str() {
+                Ok(string) => string.to_string(),
+                Err(err) => {
+                    return Err(RhinoError::new(
+                        RhinoErrorStatus::LibraryLoadError,
+                        &format!("Failed to get version info from Rhino Library: {}", err),
+                    ))
+                }
+            };
+
+            let context_info_ptr = std::ptr::null();
+            let status = pv_rhino_context_info(crhino, addr_of!(context_info_ptr));
+            check_fn_call_status!(status, "pv_rhino_context_info");
+            let context_info = CStr::from_ptr(context_info_ptr);
+
+            // Using the raw symbols means we have to ensure that "lib" outlives these refrences
+            let vtable = RhinoInnerVTable {
+                pv_rhino_process: pv_rhino_process.into_raw(),
+                pv_rhino_delete: pv_rhino_delete.into_raw(),
+                pv_rhino_is_understood: pv_rhino_is_understood.into_raw(),
+                pv_rhino_get_intent: pv_rhino_get_intent.into_raw(),
+                pv_rhino_free_slots_and_values: pv_rhino_free_slots_and_values.into_raw(),
+                pv_rhino_reset: pv_rhino_reset.into_raw(),
+            };
+
+            return Ok(Self {
+                crhino,
+                _lib: lib,
+                sample_rate,
+                frame_length,
+                version,
+                context_info: context_info.to_string_lossy().to_string(),
+                vtable,
+            });
+        }
+    }
+
+    pub fn process(&self, pcm: &[i16]) -> Result<bool, RhinoError> {
+        if pcm.len() as i32 != self.frame_length {
+            return Err(RhinoError::new(
+                RhinoErrorStatus::FrameLengthError,
+                &format!(
+                    "Found a frame length of {} Expected {}",
+                    pcm.len(),
+                    self.frame_length
+                ),
+            ));
+        }
+
+        let mut is_finalized: bool = false;
```

Comment:
though it's a bit annoying, it's good to also have a private is_finalized variable that reflects the latest return value from the process function. That way, if they call get_inference before it's ready, we can check the private variable and hand them back a rust error if they've called it too early. Otherwise, we're asking the library to handle a premature call to get_inference, which we want to avoid.

Reply:
Is there a way to poll the `C` library for this value instead?

It's not a big deal but if we have to modify a private state variable then I will need to use a mutable reference to `self` in the function which could add some extra hassle to the user. Again not a huge deal but some things just play much nicer in Rust when they are immutable.

Reply:
I don't think you can poll the c library for that (at least I've never done it). If it makes the API uglier or puts more onus on the user, don't worry about it. In frameworks where it's easy, we do this just for extra friendly error handling. The library will return an INVALID_STATE error if it gets called early, so it's not like there will be a hard crash or anything.

Reply:
In that case I will leave it as is, I tried implementing it and it does both make the API uglier and put more onus on the user.